src/sapling/saplingscriptpubkeyman.cpp

```diff
@@ -616,22 +630,24 @@ Optional<std::pair<
 
 CAmount SaplingScriptPubKeyMan::TryToRecoverAndSetAmount(const CWalletTx& tx, const SaplingOutPoint& op)
 {
-    CAmount nCredit = 0;
+    CAmount noteValue = 0;
     // if amount was not set, let's try to decrypt the note and set it.
     auto noteAndAddress = tx.DecryptSaplingNote(op);
     if (noteAndAddress) {
         const libzcash::SaplingNotePlaintext &note = noteAndAddress->first;
-        nCredit = note.value();
-        // if it's not set, then set it.
-        wallet->mapWallet[tx.GetHash()].mapSaplingNoteData[op].amount = nCredit;
+        noteValue = note.value();
     } else {
         // if cannot be decrypted, use RecoverSaplingNote.
         auto optNotePlainAndAddress = TryToRecoverNote(tx, op);
         if (optNotePlainAndAddress) {
-            nCredit += optNotePlainAndAddress->first.value();
+            noteValue = optNotePlainAndAddress->first.value();
+        } else {
+            return 0;
         }
     }
-    return nCredit;
+    // if it's not set, then set it.
```

Comment:
this isn't true, if this function is called, it's always set (while it can be decrypted), independently whether the note value is cached or note. The guard is done in the caller side.